#!/bin/bash

### Description ###

# count features from multiple BAM files using feature counts to produce a TSV count table file.

### Input variables ###

if [ $# -ne 5 ]
then
    echo "Usage: $0 <Folder containing BAM files> <GTF file> <feature to count> <output path> <output prefix>"
    echo "1) The path to the folder with split bam files as generated by split_bam.sh"
    echo "2) genome reference file (gtf)"
    echo "3) Feature to count, e.g. three_prime_utr_or_mitogene or gene, etc."
    echo "4) outpath"
    echo "5) outprefix for output files"
    exit
fi

p2splitbam=$1 
gtf=$2 # /exports/ana-scarlab/group_references/ensembl/human/102/Homo_sapiens.GRCh38.102.minusINS-IGF2_uniq3utr_MTgenes_Featurecount-annotation.gtf 
feature=$3 # 'three_prime_utr_or_mitogene'
outpath=$4
outprefix=$5
p2splitbam=${p2splitbam%/}
outpath=${outpath%/}

### Paths to utilities and scripts ###
p2featureCounts=$(which featureCounts)

### Code ###

# make output folder
mkdir -p $outpath

outfile=${outpath}/${outprefix}_count-table.txt # output from featurecounts
outtsv=${outpath}/${outprefix}_count-matrix.tsv # cleaned output

# count features in a BAM file
ls ${p2splitbam}/*.bam | \
    awk '{printf "%s", $0 " "} END {print ""}' > ${p2splitbam}/${outprefix}_cells2count.txt; c2c=${p2splitbam}/${outprefix}_cells2count.txt;

${p2featureCounts} -a $gtf\
    -o ${outfile} \
    -t $feature \
    -g gene_name \
    `cat $c2c`

# generate a tab seperated count matrix
cat ${outfile} | \
    grep -v "#" | \
    awk -F "\t|;" '{print $1 ">" $2 "\t" $0}' |\
    cut -f1,8- | sed "s,${p2splitbam}/,,g" |\
    sed "s/.bam//g" > ${outtsv}

exit