#!/bin/bash

### Description ###

# Compute miRanda binding scores for a set of mature miRNAs and eQTL variants as target sites.

### Input variables ###

if [ $# -ne 5 ]
then
    echo "Usage: $0 <eQTL CSV file> <variant VCF file> <GTF BED file> <Genome FASTA file> <output path> <output prefix>"
    echo "  1) input reference eQTL FASTA file (<outprefix>.eQTLs.REFERENCE.fasta) as generated by eqtl_to_fasta.sh"
    echo "  2) input alternative eQTL FASTA file (<outprefix>.eQTLs.ALTERNATIVE.fasta) as generated by eqtl_to_fasta.sh"
    echo "  3) input mature miRNA FASTA file downloaded from miRBase"
    echo "  4) output path"
    echo "  5) outprefix for output files"
    exit 1
fi

eqtl_ref_fasta_file=$1 
eqtl_alt_fasta_file=$2 
mature_mirna_fasta_file=$3
output_path=$4
outprefix=$5

mkdir -p ${output_path}

outprefix=${output_path%/}/${outprefix}

### Parameters ###
score_threshold=50
energy_threshold=1
scale_Z=4
gap_open=-4
gap_extend=-9

### Paths to utilities and scripts###
p2miranda=$(which miranda)

### Code ###

# Merge the reference and alternative FASTA file into one file
target_fasta_file=${eqtl_ref_fasta_file%.REFERENCE.fasta}.ALT_REF.fasta
cat ${eqtl_ref_fasta_file} > ${target_fasta_file}
cat ${eqtl_alt_fasta_file} >> ${target_fasta_file}

# Compute the alignment and energy scores for each miRNA and eQTL pair
miranda_out_file=${outprefix}.miranda.out.txt
echo "start generating ${miranda_out_file}"
${p2miranda} ${mature_mirna_fasta_file} ${target_fasta_file} -sc ${score_threshold} -en ${energy_threshold} -scale ${scale_Z} -go ${gap_open} -ge ${gap_extend} -out ${miranda_out_file}

# export data to CSV file
miranda_csv_file=${outprefix}.miranda.out.csv
cat ${miranda_out_file} |\
 grep -A 1 "Scores for this hit:" |\
 grep -v "Scores for this hit:" |\
 grep -v "^--$" |\
 awk -F "\t" '{OFS=","}{print $1,$2,$3,$4,$5, $6, $7, $8, $9}' > ${miranda_csv_file}

echo "${miranda_csv_file} made"

exit
