#!/bin/bash

### Description ###

# This script alignes a single fastq file to a reference genome using STAR to generate a BAM file 
# which is sorted and filtered by samtools.


### Input variables ###

if [ $# -ne 4 ]
then
    echo "Usage: $0 <input fastq file> <output prefix> <sampleID> <indexed genome folder (star)>"
    echo "  1) input fastq file (single file, read 1) as generated by genome_alignment.sh"
    echo "  2) outprefix for output files"
    echo "  3) Sample name to add to the read group (@RG) of the generated BAM file"
    echo "  4) genome folder containing STAR genome indices"
    exit 1
fi

inputfq=$1
outprefix=$2
samplename=$3
genome=$4


### Parameters ###
threads=8  # Adjust this if needed

### Paths to utilities ###
p2star=$(which STAR)
p2samtools=$(which samtools)
p2java=$(which java)
p2picard=/exports/ana-scarlab/bin/picard-2.25.0/picard.jar

### Code ###

# Map single-end FASTQ reads to a reference genome using STAR

${p2star} \
    --runThreadN ${threads} \
    --genomeDir ${genome} \
    --readFilesIn ${inputfq} \
    --readFilesCommand zcat \
    --outFilterMultimapNmax 20 \
    --outSAMmapqUnique 60 \
    --outSAMunmapped Within \
    --outSAMtype BAM Unsorted \
    --outSAMattributes All \
    --outFileNamePrefix ${outprefix}

# Remove temporary STAR folders
rm -r ${outprefix}_STARtmp

# Sort BAM file using samtools
${p2samtools} sort -o ${outprefix}Aligned.sortedByCoord.out.bam -@ ${threads} ${outprefix}Aligned.out.bam

# Filter BAM file using samtools
${p2samtools} view -q 60 ${outprefix}Aligned.sortedByCoord.out.bam -b -o ${outprefix}Aligned.sortedByCoord.filtered.bam

# Add read group to BAM file
${p2java} -jar ${p2picard} AddOrReplaceReadGroups \
    I=${outprefix}Aligned.sortedByCoord.filtered.bam \
    O=${outprefix}.Aligned.sortedByCoord.filtered.RG.bam RGID=${samplename} \
    RGLB=${samplename} RGPL=ILLUMINA RGPU=${samplename} \
    RGSM=${samplename} \
    USE_JDK_DEFLATER=true
