#!/bin/bash

### Description ###

# script to filter VCF files by 3'UTR variants and annotate with corresponding gene and SNP ID
# Two genes are appeneded if variant falls into the 3'UTR of 2 genes.

### Input variables ###
if [ $# -ne 5 ]
then
    echo "Usage: $0 <variant VCF file> <BED 3'UTR annotation file> <dbSNP VCF> <outprefix> <output path>"
    echo "  1) Unfiltered cohort VCF file as generated by genotype_variants.sh"
    echo "  2) BED file containing custom 3'UTR annotations"
    echo "  3) dbSNP feature file (.vcf/.vcf.gz)"
    echo "  4) outprefix"
    echo "  5) output path"
    exit 1
fi

VCFin=$1 #unfiltered cohort VCF file
three_prime_utr_BED=$2 # e.g. Homo_sapiens.GRCh38.102.minusINS-IGF2_uniq3utr_fromstop.bed 
dbsnp=$3
outprefix=$4
outdir=$5

outdir=${outdir%/}
VCF3UTR=${outdir}/${outprefix}_3UTR.vcf
VCFtmp=${VCF3UTR%.vcf}_gene_annotated.tmp.vcf
VCFout=${VCF3UTR%.vcf}_gene_annotated.vcf

### Paths to utilities ###
p2bedtools=$(which bedtools)
p2bgzip=$(which bgzip)
p2tabix=$(which tabix)
p2vcf_annotate=$(which vcf-annotate)
p2bcftools=$(which bcftools)

### Code ###

mkdir -p ${outdir}

# Select only 3'UTR variants

${p2bedtools} intersect \
	-a ${VCFin} \
	-b ${three_prime_utr_BED} \
	-header > ${VCF3UTR} 

# make gene annotation file to appened gene name
cat ${three_prime_utr_BED} |\
	awk '{OFS="\t"}BEGIN{print "#CHR\tFROM\tTO\tANNOTATION"}{print $1, $2, $3, $4}' |\
	${p2bgzip} > ${outdir}/${outprefix}.VCF.annotation.gz # compress

${p2tabix} -p bed ${outdir}/${outprefix}.VCF.annotation.gz # index

# annotate VCF file with corresponding gene
cat ${VCF3UTR} |\
	${p2vcf_annotate} -a ${outdir}/${outprefix}.VCF.annotation.gz \
	-d key=INFO,ID=Gene,Number=A,Type=String,Description=Gene \
	-c CHROM,FROM,TO,INFO/Gene > ${VCFtmp}

${p2bgzip} -c ${VCFtmp} > ${VCFtmp}.gz #compress 
${p2tabix} -p vcf ${VCFtmp}.gz #index

# annotate VCF file with SNP ID from dbSNP
${p2bcftools} annotate \
	-a ${dbsnp} \
	-c ID \
	-o ${VCFout} \
	${VCFtmp}.gz

rm -f ${VCFtmp}*

${p2bcftools}

exit
